src/pip/_internal/network/download.py

```diff
@@ -250,6 +253,77 @@ def _attempt_resumes_or_redownloads(
             os.remove(download.output_file.name)
             raise IncompleteDownloadError(download)
 
+        # If we successfully completed the download via resume, manually cache it
+        # as a complete response to enable future caching
+        if download.reattempts > 0:
+            self._cache_resumed_download(download, first_resp)
+
+    def _cache_resumed_download(
+        self, download: _FileDownload, original_response: Response
+    ) -> None:
+        """
+        Manually cache a file that was successfully downloaded via resume retries.
+
+        cachecontrol doesn't cache 206 (Partial Content) responses, since they
+        are not complete files. This method manually adds the final file to the
+        cache as though it was downloaded in a single request, so that future
+        requests can use the cache.
+        """
+        url = download.link.url_without_fragment
+        if url.startswith("https://"):
+            adapter = self._session.adapters["https://"]
+        elif url.startswith("http://"):
+            adapter = self._session.adapters["http://"]
+        else:
+            return
```

Comment:
```suggestion
        adapter = self._session.get_adapter(url)
```

This will ensure HTTP urls that are cacheable due to being marked as a trusted host are cached as well.

Reply:
Applied